### DISCUSSION TITLE
API for clearing/removing cached images

### DISCUSSION THREAD

**Discussion by ronaldbarendse** [SixLabors/ImageSharp.Web#237] [Ideas] [answered]
The current `IImageCache` API and LRU cache within the middleware don't allow clearing/removing cached images. Manually clearing the cached files while they are in the LRU cache even results in exceptions for any subsequent requests.

Cleaning up cached images might be required for legal purposes (takedowns, privacy violations, etc.) or simply for reducing storage costs.

I would suggest adding a `Clear()` method to the `IImageCache` API that accepts additional predicates, most likely based on the `ImageCacheMetadata`, so you can clear files:
- of a specific content type
- source or cached files created in the last 5 minutes or over 1 month ago
- expired cache files (older than the max age)
- smaller/larger than a specific size

The LRU cache should then obviously retry generating the image, instead of throwing an exception when the cached file can't be found anymore. Any thoughts on this before shipping v2 (as this will be a breaking change).

**Comment by JimBobSquarePants** [comment, marked as answer]
The problem with a clear option is that it’s potentially expensive and should be non-blocking. That’s why clearing was always left as an implementation detail of whatever site is using the middleware. I don’t think a reasonable implementation can be added. 

If someone wants a clearable cache I would suggest they implement the cache interface and use their own logic

**Reply by ronaldbarendse** [reply]
So basically the only safe way to clean up the cache is to shut down the application first? 🤔

I would be OK if the implementation would be something you implement yourself (with the caveat that you don't end up trying to delete files that are being written/read at the same time). Currently this won't work though, because you can't clear the internal LRU cache... So maybe that `Clear()` method should actually be added to the middleware instead?

**Reply by JimBobSquarePants** [reply]
We should fix whatever issue is causing the LRU cache to throw if a cache item is returned that is then null. It's not supposed to do that.

Edit. A quick look at the code suggests that the error is not thrown in the cache but later in `SendResponseAsync`. I could be wrong though. I'll do a thorough investigation and add a fix now.

**Reply by ronaldbarendse** [reply]
Awesome to already see this being fixed in https://github.com/SixLabors/ImageSharp.Web/pull/240 🙌🏻
